spec/p2p/v0.34/README.md

```diff
@@ -0,0 +1,140 @@
+| Peer ranking | Connection upgrading | Evicting| 
+| --- | --- | --- |
+| PEX / reactors (only marking peers as good/bad) | - | Switch| 
+
+
+The main states a node can be in are:
+- 'Candidate': peers that we have not connected to yet but are discovered. (no explicit list)
+- 'Dialing' : peers who are currently being dialed. ( I did not see where this is of particular use. )
+- 'Peers' : Connected peers (effectivley a connected state)
+- 'Reconnecting': Peers to whom a node is currently reconnecting. The node is trying to establosh a connectiong to these peers and is failing to do so. After a certain amount of time, the peer is simply dropped to be discovered again by the PEX reactor. 
+- 'BadPeers' : This is the only list not kept by the switch. It is stored within the address book of the PEX reactor. 
+
+The diagram below shows the lifecycle of an outbound peer:
+
+<img src="img/p2p_state.png" width="50%" title="Outgoing peers lifecycle">Outgoing peers lifecycle</img>
+
+
+### Peer discovery
+
+Peers are discovered by adding addresses provided in the config file or triggering dials to seed nodes or nodes in the address book. 
+
+When a node is started, it provides the list of persistent peers to the switch by calling `DialPeersAsync`.
+
+Depending on whether the node is a seed node or or not, the PEX reactor constantly runs either `crawlPeersRoutine` or `ensurePeers()` respectively. 
+If the node is a seed node, `crawlPeersRoutine` reads peer information randomly from the address book, tries to dial the peer and requests from them information on other peers.
+
+When a node receives information about other peers from a seed note, it sends another request to the node asking for more peers. 
+
+A node also learns of other peers when they try to connect to it. 
+
+### Dialing peers
+
+Every node connects to a set of peers to whom it sends requests via the Switch. 
+
+The PEX reactor ensures a node is connected to peers by running a routine to keep dialing peers in the background until a threshold of connected peers is reached (`MaxNumOutboundPeers`). This way, if a peers' address has been added via an incoming request, it will eventually be dialed. Peers to dial are chosen with some configurable bias for unvetted peers. The bias should be lower when we have fewer peers and can increase as we obtain more, ensuring that our first peers are more trustworthy, but always giving us the chance to discover new good peers. 
+
+As the number of outgoing peers is limited, the reactor will choose the number of addresses to dial taking into account the current number of outgoing connections and ongoing dialings. The addresses to dial will be picked with a bias towards new and vetted peers (TODO define ).
+
+Except for persistent peers, all other peers can be dialed at most `maxAttemptsToDial`. If a node is not connected by that time it is marked as bad. Otherwise, if the dial fails, the node will wait before the next dial. (exponential backoff mechanism). 
```

Comment:
There is a linear backoff, followed by an exponential backoff. See `reconnectToPeer` method in the switch.